rotkehlchen/chain/ethereum/modules/oneinch/v4/decoder.py

```diff
@@ -0,0 +1,85 @@
+from typing import TYPE_CHECKING, Callable
+
+from rotkehlchen.accounting.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.chain.evm.decoding.oneinch.decoder import OneinchCommonDecoder
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+from rotkehlchen.types import ChecksumEvmAddress, EvmTransaction
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.accounting.structures.evm_event import EvmEvent
+
+from rotkehlchen.chain.ethereum.modules.uniswap.v3.decoder import (
+    SWAP_SIGNATURE as UNISWAP_V3_SWAP_SIGNATURE,
+)
+
+from ..constants import CPT_ONEINCH_V4
+from .constants import ONEINCH_V4_MAINNET_ROUTER
+
+
+class Oneinchv4Decoder(OneinchCommonDecoder):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+            router_address=ONEINCH_V4_MAINNET_ROUTER,
+            swapped_signature=b'v\xaf"J\x148e\xa5\x0bAIn\x1asb&\x98i,V\\\x12\x14\xbc\x86/\x18\xe2-\x82\x9c^',  # noqa: E501
+            counterparty=CPT_ONEINCH_V4,
+        )
+
+    def _handle_post_decoding(
+            self,
+            transaction: EvmTransaction,
+            decoded_events: list['EvmEvent'],
+            all_logs: list[EvmTxReceiptLog],
+    ) -> list['EvmEvent']:
+        """
+        It handles the case where a swap is done on Uniswap v3 via 1inch v4. In that case
+        there are 3 simple decoded events already created: Fee, spend and receive. Spend and
+        receive events are not decoded by the uniswap v3 decoder and should be decoded by this
+        aggregator decoder. This function updates the spend and receive events so that the swap
+        is properly decoded. (The decoding is done with this post decoding rule and not from
+        addresses_to_decoders mapping because the 1inch v4 router is not in the addresses of the
+        logs of the transaction, and consequently it can't be mapped.)
+        """
+        for tx_log in all_logs:
+            if tx_log.topics[0] == UNISWAP_V3_SWAP_SIGNATURE:
+                sender = transaction.from_address
+                for event in decoded_events:
+                    if event.event_type == HistoryEventType.RECEIVE and event.location_label == sender:  # maybe amount and asset as well?  # noqa: E501
```

Comment:
why not match on amount and asset?

Reply:
The asset is not part of the Swap event and the amount of the Swap event is not the amount of the complete swap we want to show to the user. The trasnaction might have many smaller swap events and the total amount is the total swap amount we show to teh user (also I think there are some fees by 1inch). In all transactions that I tested, this approach works fine.